compio-runtime/src/attacher.rs

```diff
@@ -53,39 +57,169 @@ impl Attacher {
-        if cfg!(windows) {
-            Ok(self.clone())
+    fn try_clone(&self) -> io::Result<Self> {
+        let source = self.source.try_clone()?;
+        let new_self = if cfg!(windows) {
+            Self {
+                source,
+                once: self.once.clone(),
+            }
         } else {
-            let new_self = Self::new();
+            let new_self = Self::new(source);
             if self.is_attached() {
-                new_self.attach(source)?;
+                new_self.attach()?;
             }
-            Ok(new_self)
-        }
+            new_self
+        };
+        Ok(new_self)
+    }
+}
+
+impl<S> IntoInner for Attacher<S> {
+    type Inner = S;
+
+    fn into_inner(self) -> Self::Inner {
+        self.source
     }
 }
 
 /// Represents an attachable resource to driver.
 pub trait Attachable {
-    /// Attach self to the global driver.
-    fn attach(&self) -> io::Result<()>;
-
     /// Check if [`Attachable::attach`] has been called.
     fn is_attached(&self) -> bool;
 }
 
+/// Duplicatable file or socket.
+pub trait TryClone: Sized {
+    /// Duplicate the source.
+    fn try_clone(&self) -> io::Result<Self>;
+}
+
+impl TryClone for std::fs::File {
+    fn try_clone(&self) -> io::Result<Self> {
+        std::fs::File::try_clone(self)
+    }
+}
+
+impl TryClone for socket2::Socket {
+    fn try_clone(&self) -> io::Result<Self> {
+        socket2::Socket::try_clone(self)
+    }
+}
+
+#[cfg(windows)]
+impl TryClone for OwnedHandle {
+    fn try_clone(&self) -> io::Result<Self> {
+        OwnedHandle::try_clone(self)
+    }
+}
+
+#[cfg(windows)]
+impl TryClone for OwnedSocket {
+    fn try_clone(&self) -> io::Result<Self> {
+        OwnedSocket::try_clone(self)
+    }
+}
+
+#[cfg(unix)]
+impl TryClone for OwnedFd {
+    fn try_clone(&self) -> io::Result<Self> {
+        OwnedFd::try_clone(self)
+    }
+}
+
+/// Extracts raw fds.
+pub trait TryAsRawFd {
+    /// Get the inner raw fd, while ensuring the source being attached.
+    fn try_as_raw_fd(&self) -> io::Result<RawFd>;
+
+    /// Get the inner raw fd and don't check if it has been attached.
+    ///
+    /// # Safety
+    ///
+    /// The caller should ensure it is attached before submit an operation with
+    /// it.
+    unsafe fn as_raw_fd_unchecked(&self) -> RawFd;
+}
+
+impl<T: AsRawFd> TryAsRawFd for T {
+    fn try_as_raw_fd(&self) -> io::Result<RawFd> {
+        Ok(self.as_raw_fd())
+    }
+
+    unsafe fn as_raw_fd_unchecked(&self) -> RawFd {
+        self.as_raw_fd()
+    }
+}
+
+impl<S: AsRawFd> TryAsRawFd for Attacher<S> {
+    fn try_as_raw_fd(&self) -> io::Result<RawFd> {
+        Ok(self.try_get()?.as_raw_fd())
+    }
+
+    unsafe fn as_raw_fd_unchecked(&self) -> RawFd {
+        self.source.as_raw_fd()
+    }
+}
+
 /// A [`Send`] wrapper for attachable resource that has not been attached. The
 /// resource should be able to send to another thread before attaching.
 pub struct Unattached<T: Attachable>(T);
```

Comment:
Didn't see this before but trait bounds should not be on struct's, unless necessary.

Reply:
Ok.

Reply:
Ok